---
title: "BCR-ABL1 in depth RNA Analysis"
author: "Leonardo Amato Regis de Farias"
date: "2022-08-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
PTH <- "clinvar_result.txt"
if (!file.exists(PTH)) {
  print("Recheck if your file is in the following directory:")
  print(getwd())
} else {
    clin_df <- read.delim(PTH)
}
```

```{r echo=FALSE}
#head(clin_df)
print("Here are the following features:")
print(colnames(clin_df))
dim(clin_df)
```

#### we need to find about the uniqueness of the features. For instance, just after tested examination on __, there was only ONE unique value, chromosome 9, so no purpose in keeping the feature
```{r}
clin_mat <- as.matrix(clin_df)  #to support idx-accessible computation
high_bounds <- c()
for (feature in colnames(clin_df)) {
  nunique = length(unique(clin_mat[, feature]))
  print( paste("For this feature:", feature, "---> there are", nunique, "unique values") )
  if (nunique==1 || nunique > 200) {
    #either too much diversity that gives no correlation, or literally just one value
    high_bounds <- append(high_bounds, feature)
  }
}
```
##### just by observing above, 'X' and 'GRCh37Chromosome' are completely useless... With 227 values, many features possess high variations.

```{r echo=FALSE}
useless <- clin_mat[, high_bounds]
clin_final <- clin_df[, -match(high_bounds, names(clin_df))] 
clin_mat <- as.matrix(clin_final)
```

##### check if there are any missing values
```{r}
sum(is.na(clin_final))  #25 missing in total
na.s <- c()
for (feature in 1:9) {
  amt_na <- sum(is.na(clin_mat[,feature]))
  na.s <- append(na.s, amt_na )
  if (amt_na > 0) {
    print(paste(colnames(clin_final)[feature], "has some missing"))
  }
}
```

#### why does 'GRCh38Chromosome' have some null values?
#### its significance: "GRCh38/hg38 is the assembly of the human genome released December of 2013" (https://gatk.broadinstitute.org/hc/en-us/articles/360035890951-Human-genome-reference-builds-GRCh38-or-hg38-b37-hg19)
```{r}
#integer datatype
class(clin_final$GRCh38Chromosome)

#doesn't quite modify the values because R is weird on returns
imputer <- function(col_vector, method="mean") {
  fillna <- NaN
  if (method == "mean" || method=="average") {
    fillna <- mean(col_vector)
  }
  #probably not the best choice for integer column vector
  else if (method == "median") {
    fillna <- median(col_vector)
  }
  for (ridx in which(is.na(col_vector)) ) {
    col_vector[ridx] <- fillna
  }
  col_vector <- col_vector
}
#clin_final$GRCh38Chromosome <- imputer(clin_final$GRCh38Chromosome)
```

#### either exclude GRCh38Chromosome altogether or impute with zeroes
```{r}
#using the first option, total exclusion
clin_final <- subset(clin_final, select = -c(GRCh38Chromosome))
head(clin_final[, 2:8], 3)  #just annoying to see empty space made by the first feature
```

```{r}
#probably unecessary to utilize the other alternative
```

```{r}
#none of the features are numeric
clin_mat <- as.matrix(clin_final)
for (feature in colnames(clin_final)) {
  print( paste(feature, class(clin_mat[feature]), sep=": ") )
}
```

#### below are the possible classes for patients
```{r}
conditions <- unique(clin_final$Condition.s.)
#... <- as.factor(conditions)
severity <- as.factor(clin_final$Clinical_significance)
toPred <- c(5, 13, 14, 15, 18, 19, 20, 22)
conditions
```

#### what will be a complex challenge is handling the 'Gene.s.' attribute, which is separated by | for each row sample. ABL1 is prevalent in many of the samples.
```{r}
genes <- clin_final$Gene.s.
freq <- c()  #entire number of genes summed up
condition_freq <- c()  #also need to record conditions for genes

for (pos in 1:length(genes) ) {
  sample = genes[pos]
  genes_per_samples <- strsplit(sample, split="|", fixed=TRUE)
  
  patient_condition <- clin_final[pos, "Condition.s."]
  if (!( patient_condition %in% names(condition_freq) )) {
    condition_freq[patient_condition] = c()  #another hashmap/dictionary data structure
  }
  
  for (G in genes_per_samples[[1]]) {
    #again, counting number of genes in general
    if (!(G %in% names(freq))) {
      freq[G] = 0
    }
    freq[G] = freq[G]+1
    if ( !(G %in% names(condition_freq[[patient_condition]])) ) {
      condition_freq[[patient_condition]][G] = 0
    }
    condition_freq[[patient_condition]][G] = condition_freq[[patient_condition]][G]+1
  }
}

###need to be able to count the amount each gene per leukemia type
condition_freq[toPred]
```

```{r}
barplot(freq, main="The frequency of occurances per gene for this dataset")
```

#### encoding the genetic attributes with ordinal and label encoders, but also normalizing some because of their biological significance
```{r}
#extremely efficient pattern/ordinal encoder I created
patternEnc <- function(feature, patterns, data=clin_final, LE=0) {
  original <- as.matrix(data[,feature])
  encoded <- rep(0, nrow(data))
  for (mod_ in patterns) {
    for (j in grep(mod_, original)) {
      encoded[j] = LE
    }
    LE = LE + 1
  }
  encoded
}

levels <- c("not provided", "Uncertain", "Benign", "Likely benign", "Likely Patho", "Pathogenic")
clin_final$levels <- patternEnc("Clinical_significance", levels, data=clin_final, LE=-1)

review <- unique(clin_final$Review_status)
clin_final$new_Review_status <- patternEnc("Review_status", review, data=clin_final)

conditionEnc <- unique(clin_final$Condition.s.)
clin_final$new_Condition.s. <- patternEnc("Condition.s.", conditionEnc, data=clin_final)

###fix
#more specific to whether degree of leukemia (0 if none)
leukEnc <- grep("eukemia", unique(clin_final$Condition.s.))
clin_final$leukemia <- patternEnc("Condition.s.", c("None",unique(clin_final$Condition.s.)[leukEnc]))
```

```{r}
clin_final$leukemia
```









































